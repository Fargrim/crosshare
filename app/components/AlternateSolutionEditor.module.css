.page {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.main {
  position: relative;
  overflow: scroll;
  flex: 1 1 auto;
}

.container {
  position: absolute;

  container-type: size;
  display: flex;
  flex: 1 1 auto;
  flex-flow: column nowrap;
  align-items: center;

  width: 100%;
  height: 100%;

  outline: none;
}
