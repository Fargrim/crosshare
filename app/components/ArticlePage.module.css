@value huge-and-up from '../lib/definitions.module.css';
@value max-width from '../lib/definitions.module.css';

.page {
  margin: 2em 1em;
}

@media (huge-and-up) {
  .page {
    max-width: max-width;
    margin: 2em auto;
  }
}

.markdown {
  margin-bottom: 2em;
}

.markdown h2 {
  margin-top: 1em;
}
