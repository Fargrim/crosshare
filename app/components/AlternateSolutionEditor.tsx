import useEventListener from '@use-it/event-listener';
import {
  useCallback,
  useReducer,
  useMemo,
  useRef,
  useState,
  useEffect,
} from 'react';
import {
  FaSave,
  FaWindowClose,
  FaEllipsisH,
  FaVolumeUp,
  FaVolumeMute,
  FaKeyboard,
} from 'react-icons/fa';
import { usePersistedBoolean, usePolyfilledResizeObserver } from '../lib/hooks';
import {
  Direction,
  fromKeyboardEvent,
  fromKeyString,
  KeyK,
} from '../lib/types';
import { fromCells } from '../lib/viewableGrid';
import { Rebus, EscapeKey } from './Icons';
import { Keyboard } from './Keyboard';
import {
  TopBar,
  TopBarLink,
  TopBarDropDown,
  TopBarDropDownLink,
} from './TopBar';
import {
  gridInterfaceReducer,
  KeypressAction,
  PasteAction,
} from '../reducers/reducer';
import { isSome } from 'fp-ts/lib/Option';
import { GridView } from './Grid';
import { logAsyncErrors } from '../lib/utils';
import { isTextInput } from '../lib/domUtils';

export function AlternateSolutionEditor(props: {
  grid: string[];
  width: number;
  height: number;
  vBars: Set<number>;
  hBars: Set<number>;
  hidden: Set<number>;
  highlighted: Set<number>;
  highlight: 'circle' | 'shade';
  cancel: () => void;
  save: (alt: Record<number, string>) => Promise<void>;
}) {
  const initialGrid = fromCells({
    mapper: (e) => e,
    width: props.width,
    height: props.height,
    cells: props.grid,
    vBars: props.vBars,
    hBars: props.hBars,
    allowBlockEditing: false,
    highlighted: props.highlighted,
    highlight: props.highlight,
    hidden: props.hidden,
  });

  const [state, dispatch] = useReducer(gridInterfaceReducer, {
    type: 'alternate-editor',
    active: { col: 0, row: 0, dir: Direction.Across },
    grid: initialGrid,
    wasEntryClick: false,
    showExtraKeyLayout: false,
    isEnteringRebus: false,
    rebusValue: '',
    downsOnly: false,
    isEditable: () => true,
  });

  const [muted, setMuted] = usePersistedBoolean('muted', true);
  const [toggleKeyboard, setToggleKeyboard] = usePersistedBoolean(
    'keyboard',
    false
  );

  const physicalKeyboardHandler = useCallback(
    (e: KeyboardEvent) => {
      const mkey = fromKeyboardEvent(e);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
        e.preventDefault();
      }
    },
    [dispatch]
  );
  useEventListener('keydown', physicalKeyboardHandler);

  const containerRef = useRef<HTMLDivElement | null>(null);
  const { width: cqw, height: cqh } = usePolyfilledResizeObserver(containerRef);
  const [useCQ, setUseCQ] = useState(true);
  useEffect(() => {
    if (!('container' in document.documentElement.style)) {
      setUseCQ(false);
    }
  }, []);

  const pasteHandler = useCallback(
    (e: ClipboardEvent) => {
      if (isTextInput(e.target)) {
        return;
      }
      const pa: PasteAction = {
        type: 'PASTE',
        content: e.clipboardData?.getData('Text') ?? '',
      };
      dispatch(pa);
      e.preventDefault();
    },
    [dispatch]
  );
  useEventListener('paste', pasteHandler);

  const keyboardHandler = useCallback(
    (key: string) => {
      const mkey = fromKeyString(key);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
      }
    },
    [dispatch]
  );

  const topBarChildren = useMemo(() => {
    return (
      <>
        <TopBarLink
          icon={<FaSave />}
          text="Add Alternate"
          onClick={logAsyncErrors(async () => {
            const alt: Record<number, string> = {};
            let hadAny = false;
            for (const [idx, cellValue] of state.grid.cells.entries()) {
              const defaultCellValue = initialGrid.cells[idx];
              if (
                cellValue.trim() &&
                cellValue.trim() != defaultCellValue?.trim()
              ) {
                hadAny = true;
                alt[idx] = cellValue;
              }
            }
            if (!hadAny) {
              props.cancel();
              return;
            }
            return props.save(alt).then(() => {
              props.cancel();
            });
          })}
        />
        <TopBarLink
          icon={<FaWindowClose />}
          text="Cancel"
          onClick={props.cancel}
        />
        <TopBarDropDown icon={<FaEllipsisH />} text="More">
          {() => (
            <>
              <TopBarDropDownLink
                icon={<Rebus />}
                text="Enter Rebus"
                shortcutHint={<EscapeKey />}
                onClick={() => {
                  const a: KeypressAction = {
                    type: 'KEYPRESS',
                    key: { k: KeyK.Escape },
                  };
                  dispatch(a);
                }}
              />
              {muted ? (
                <TopBarDropDownLink
                  icon={<FaVolumeUp />}
                  text="Unmute"
                  onClick={() => {
                    setMuted(false);
                  }}
                />
              ) : (
                <TopBarDropDownLink
                  icon={<FaVolumeMute />}
                  text="Mute"
                  onClick={() => {
                    setMuted(true);
                  }}
                />
              )}
              <TopBarDropDownLink
                icon={<FaKeyboard />}
                text="Toggle Keyboard"
                onClick={() => {
                  setToggleKeyboard(!toggleKeyboard);
                }}
              />
            </>
          )}
        </TopBarDropDown>
      </>
    );
  }, [
    initialGrid.cells,
    muted,
    props,
    setMuted,
    setToggleKeyboard,
    state.grid.cells,
    toggleKeyboard,
  ]);

  const aspectRatio = props.width / props.height;

  return (
    <>
      <div
        css={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%',
        }}
      >
        <div css={{ flex: 'none' }}>
          <TopBar>{topBarChildren}</TopBar>
        </div>
        <div
          css={{ flex: '1 1 auto', overflow: 'scroll', position: 'relative' }}
        >
          <div
            // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
            tabIndex={0}
            css={{
              outline: 'none',
              display: 'flex',
              flex: '1 1 auto',
              flexDirection: 'column',
              alignItems: 'center',
              height: '100%',
              width: '100%',
              position: 'absolute',
              flexWrap: 'nowrap',
              containerType: 'size',
            }}
            ref={containerRef}
          >
            <div
              aria-label="grid"
              css={{
                margin: 'auto',
                width: useCQ
                  ? `min(100cqw, 100cqh * ${aspectRatio})`
                  : `min(${cqw}px, ${cqh}px * ${aspectRatio})`,
                height: useCQ
                  ? `min(100cqh, 100cqw / ${aspectRatio})`
                  : `min(${cqh}px, ${cqw}px / ${aspectRatio})`,
              }}
            >
              <GridView
                isEnteringRebus={state.isEnteringRebus}
                rebusValue={state.rebusValue}
                grid={state.grid}
                defaultGrid={initialGrid}
                active={state.active}
                dispatch={dispatch}
                allowBlockEditing={false}
                autofill={[]}
              />
            </div>
          </div>
        </div>
        <div css={{ flex: 'none', width: '100%' }}>
          <Keyboard
            toggleKeyboard={toggleKeyboard}
            keyboardHandler={keyboardHandler}
            muted={muted}
            showExtraKeyLayout={state.showExtraKeyLayout}
            includeBlockKey={false}
          />
        </div>
      </div>
    </>
  );
}
