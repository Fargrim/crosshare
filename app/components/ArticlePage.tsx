import type { Root } from 'hast';
import Head from 'next/head';
import { ArticleT } from '../lib/article';
import { ArticlePageProps } from '../lib/serverOnly';
import { HUGE_AND_UP, MAX_WIDTH } from '../lib/style';
import { ContactLinks } from './ContactLinks';
import { ErrorPage } from './ErrorPage';
import { Link } from './Link';
import { Markdown } from './Markdown';
import { DefaultTopBar } from './TopBar';

export function ArticlePage(props: ArticlePageProps) {
  if ('error' in props) {
    return (
      <ErrorPage title="Article Not Found">
        <p>
          We&apos;re sorry, we couldn&apos;t find the article you requested.
        </p>
        <p>{props.error}</p>
        <p>
          Try the <Link href="/">homepage</Link>.
        </p>
      </ErrorPage>
    );
  }
  return <Article key={props.s} {...props} />;
}

function Article(props: ArticleT & { hast: Root }) {
  return (
    <>
      <Head>
        <title>{props.t}</title>
        <meta key="og:title" property="og:title" content={props.t} />
        <link
          rel="canonical"
          href={'https://crosshare.org/article/' + props.s}
        />
      </Head>
      <DefaultTopBar />
      <div
        css={{
          margin: '2em 1em',
          [HUGE_AND_UP]: {
            maxWidth: MAX_WIDTH,
            margin: '2em auto',
          },
        }}
      >
        <Markdown
          css={{ marginBottom: '2em', '& h2': { marginTop: '1em' } }}
          hast={props.hast}
        />
        <p>
          This article is part of a series of posts designed to teach visitors
          about crosswords in general as well as some Crosshare specific
          features. If you have any questions or suggestions for this or other
          articles please contact us via <ContactLinks />.
        </p>
        <p>
          We&apos;re seeking volunteers to help expand and edit this
          knowledgebase so it becomes more useful for constructors and solvers.
          If you&apos;re interested please reach out!
        </p>
      </div>
    </>
  );
}
