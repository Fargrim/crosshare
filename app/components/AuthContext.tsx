import { ReactNode, createContext, useContext } from 'react';
import { ErrorPage } from '../components/ErrorPage';
import type { User } from 'firebase/auth';
import { GoogleLinkButton, GoogleSignInButton } from './GoogleButtons';
import { TopBar } from './TopBar';
import { ConstructorPageT } from '../lib/constructorPage';
import { NotificationT } from '../lib/notificationTypes';
import { AccountPrefsT } from '../lib/prefs';
import { WithConditionalCSSProp } from '@emotion/react/types/jsx-namespace';

export interface AuthProps {
  isAdmin: boolean;
  user: User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export interface AuthPropsOptional {
  isAdmin: boolean;
  user?: User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export function renderLoginButtonIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return <GoogleSignInButton />;
  }
  if (user.isAnonymous) {
    return <GoogleLinkButton user={user} />;
  }
  return null;
}

function renderLoginIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleSignInButton />
        </div>
      </>
    );
  }
  if (user.isAnonymous) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleLinkButton user={user} />
        </div>
      </>
    );
  }
  return null;
}

/* Ensure we have a non-anonymous user, upgrading an anonymous user if we have one. */
export function requiresAuth<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={ctx.isAdmin}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

/* Ensure we have an admin user, upgrading an anonymous user if we have one. */
export function requiresAdmin<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    if (!ctx.isAdmin) {
      return (
        <ErrorPage title="Not Allowed">
          <p>You do not have permission to view this page</p>
        </ErrorPage>
      );
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={true}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

export interface AuthContextValue {
  user?: User;
  notifications?: Array<NotificationT>;
  isAdmin: boolean;
  isPatron: boolean;
  loading: boolean;
  error?: string;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
  displayName?: string | null;
  updateDisplayName?: (n: string) => Promise<void>;
}
export const AuthContext = createContext<AuthContextValue>({
  user: undefined,
  isAdmin: false,
  isPatron: false,
  loading: false,
  error: 'using default context',
  constructorPage: undefined,
  prefs: undefined,
});
