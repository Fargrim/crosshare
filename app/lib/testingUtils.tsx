import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';

import { AuthContext } from '../components/AuthContext';
import { SnackbarProvider } from '../components/Snackbar';

import { GetServerSidePropsResult } from 'next';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { hasOwnProperty } from './types';
import type { User } from 'firebase/auth';

export const getUser = (uid: string, isAnonymous: boolean) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const u = { uid, isAnonymous } as any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  u.updateProfile = (_profile: { displayName?: string }) => {
    u.displayName = _profile.displayName;
    return Promise.resolve();
  };
  return u as User;
};

export const anonymousUser = getUser('anonymous-user-id', true);

import { messages as messagesEn } from '../locales/en/messages';

i18n.load({
  en: messagesEn,
});
i18n.activate('en');

const WithAllProviders: (
  opts: AuthOptions,
  includeSnackbar?: boolean
) => ComponentType =
  (opts: AuthOptions, includeSnackbar?: boolean) =>
  ({ children }: { children?: ReactNode }) => {
    return (
      <I18nProvider i18n={i18n}>
        <AuthContext.Provider
          value={{
            user: undefined,
            isAdmin: false,
            isPatron: false,
            loading: false,
            error: undefined,
            ...opts,
          }}
        >
          {includeSnackbar ? (
            <SnackbarProvider>{children}</SnackbarProvider>
          ) : (
            children
          )}
        </AuthContext.Provider>
      </I18nProvider>
    );
  };

interface AuthOptions {
  user?: User;
  isAdmin?: boolean;
  displayName?: string;
}

function wrappedRender(
  ui: ReactElement,
  auth: AuthOptions,
  options?: RenderOptions
): RenderResult {
  return render(ui, {
    wrapper: WithAllProviders(auth),
    ...options,
  });
}

export function renderWithSnackbar(
  ui: ReactElement,
  auth: AuthOptions,
  options?: RenderOptions
): RenderResult {
  return render(ui, {
    wrapper: WithAllProviders(auth, true),
    ...options,
  });
}

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';

// override render method
// eslint-disable-next-line import/export
export { wrappedRender as render };

export async function getProps<T>(a: GetServerSidePropsResult<T>) {
  if (!hasOwnProperty(a, 'props')) {
    throw new Error('should have props');
  }
  return await Promise.resolve(a.props);
}
