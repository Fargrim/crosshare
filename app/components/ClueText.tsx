import { useContext } from 'react';

import { CluedEntry } from '../lib/viewableGrid';

import { Direction, getClueText } from '../lib/types';
import { Markdown } from './Markdown';
import { DownsOnlyContext } from './DownsOnlyContext';

interface ClueTextProps {
  entry: CluedEntry;
}
export const ClueText = (props: ClueTextProps) => {
  const downsOnly = useContext(DownsOnlyContext);
  if (downsOnly && props.entry.direction === Direction.Across) {
    return <span>-</span>;
  }
  const text = getClueText(props.entry);
  if (props.entry.clue.startsWith('!#')) {
    return <span>{text}</span>;
  }
  let noRefs = false;
  if (props.entry.clue.startsWith('!@')) {
    console.log('setting norefs');
    noRefs = true;
  }
  return <Markdown text={text} inline={true} noRefs={noRefs} />;
};
