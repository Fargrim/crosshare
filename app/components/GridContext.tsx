import { createContext } from 'react';
import { CluedGrid } from '../lib/viewableGrid';

export const GridContext = createContext<CluedGrid | null>(null);
