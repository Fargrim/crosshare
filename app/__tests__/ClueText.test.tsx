import { act, render } from '../lib/testingUtils';
import { addClues, CluedGrid, fromCells } from '../lib/viewableGrid';
import { ClueText } from '../components/ClueText';
import { GridContext } from '../components/GridContext';

/*
"What's a Grecian ___?" "About $25 a week!"
_____ Ha'i, from South Pacific

*/

test('Highlighting for different types of clues', async () => {
  const answers = ['U', 'P', 'S', 'O'];
  const grid = fromCells({
    width: 2,
    height: 2,
    cells: answers,
    allowBlockEditing: false,
    highlighted: new Set<number>(),
    vBars: new Set<number>(),
    hBars: new Set<number>(),
    hidden: new Set<number>(),
    highlight: 'circle',
    mapper: (x) => x,
  });

  const cluedGrid: CluedGrid = addClues(grid, [
    { num: 1, dir: 0, clue: 'Not down 5-acrosses', explanation: null },
    { num: 3, dir: 0, clue: '2-down Then...', explanation: null },
    { num: 1, dir: 1, clue: '1- and 3- acrosses You and I', explanation: null },
    { num: 2, dir: 1, clue: '3A Post office abbr.', explanation: null },
  ]);

  for (let i = 0; i < 4; i += 1) {
    const entry = cluedGrid.entries[i];
    if (!entry) {
      throw new Error('oob');
    }
    const { container } = render(
      <div>
        <GridContext.Provider value={cluedGrid}>
          <ClueText entry={entry} />
        </GridContext.Provider>
      </div>,
      {}
    );
    expect(container).toMatchSnapshot();
    await act(async () => Promise.resolve()); // Popper update() - https://github.com/popperjs/react-popper/issues/350
  }

  const first = cluedGrid.entries[0];
  if (!first) {
    throw new Error('oob');
  }

  let { container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: 'Try 3-a' }} />
    </GridContext.Provider>,
    {}
  );
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: '!@ Try 3-a' }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: 'Try 3-a in **bold**!' }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: '!@ Try 3-a in **bold**!' }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: '!#Try 3-a in **bold**!' }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText
        entry={{ ...first, clue: 'Something like 3As or 1Ds should ref' }}
      />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: '#1 grossing movie of all time' }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText entry={{ ...first, clue: "_____ Ha'i, from South Pacific" }} />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();

  ({ container } = render(
    <GridContext.Provider value={cluedGrid}>
      <ClueText
        entry={{
          ...first,
          clue: '"What\'s a Grecian ___?" "About $3 a week!"',
        }}
      />
    </GridContext.Provider>,
    {}
  ));
  expect(container).toMatchSnapshot();
});
