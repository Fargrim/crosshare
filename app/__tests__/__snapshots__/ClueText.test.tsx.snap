// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighting for different types of clues 1`] = `
<div>
  <div>
    <div>
      Not down 
      5-acrosses
    </div>
  </div>
</div>
`;

exports[`Highlighting for different types of clues 2`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <div>
      <span
        class="emotion-0"
      >
        1
      </span>
      <div
        class="emotion-1"
        style="position: fixed; left: 0px; top: 0px;"
      >
        Not down 5-acrosses
        <b
          class="emotion-2"
        >
          [
          U
          P
          ]
        </b>
        <div
          class="emotion-3"
          style="position: absolute;"
        />
      </div>
      - and 
      <span
        class="emotion-0"
      >
        3- acrosses
      </span>
      <div
        class="emotion-1"
        style="position: fixed; left: 0px; top: 0px;"
      >
        2-down Then...
        <b
          class="emotion-2"
        >
          [
          S
          O
          ]
        </b>
        <div
          class="emotion-3"
          style="position: absolute;"
        />
      </div>
       You and I
    </div>
  </div>
</div>
`;

exports[`Highlighting for different types of clues 3`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <div>
      <span
        class="emotion-0"
      >
        3A
      </span>
      <div
        class="emotion-1"
        style="position: fixed; left: 0px; top: 0px;"
      >
        2-down Then...
        <b
          class="emotion-2"
        >
          [
          S
          O
          ]
        </b>
        <div
          class="emotion-3"
          style="position: absolute;"
        />
      </div>
       Post office abbr.
    </div>
  </div>
</div>
`;

exports[`Highlighting for different types of clues 4`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <div>
      <span
        class="emotion-0"
      >
        2-down
      </span>
      <div
        class="emotion-1"
        style="position: fixed; left: 0px; top: 0px;"
      >
        3A Post office abbr.
        <b
          class="emotion-2"
        >
          [
          P
          O
          ]
        </b>
        <div
          class="emotion-3"
          style="position: absolute;"
        />
      </div>
       Then...
    </div>
  </div>
</div>
`;

exports[`Highlighting for different types of clues 5`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    Try 
    <span
      class="emotion-0"
    >
      3-a
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      2-down Then...
      <b
        class="emotion-2"
      >
        [
        S
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
  </div>
</div>
`;

exports[`Highlighting for different types of clues 6`] = `
<div>
  <div>
    Try 3-a
  </div>
</div>
`;

exports[`Highlighting for different types of clues 7`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    Try 
    <span
      class="emotion-0"
    >
      3-a
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      2-down Then...
      <b
        class="emotion-2"
      >
        [
        S
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
     in 
    <strong>
      bold
    </strong>
    !
  </div>
</div>
`;

exports[`Highlighting for different types of clues 8`] = `
<div>
  <div>
    Try 3-a in 
    <strong>
      bold
    </strong>
    !
  </div>
</div>
`;

exports[`Highlighting for different types of clues 9`] = `
<div>
  <span>
    Try 3-a in **bold**!
  </span>
</div>
`;

exports[`Highlighting for different types of clues 10`] = `
<div>
  <div>
    Something like 3As or 1Ds should ref
  </div>
</div>
`;

exports[`Highlighting for different types of clues 11`] = `
<div>
  <div>
    #1 grossing movie of all time
  </div>
</div>
`;

exports[`Highlighting for different types of clues 12`] = `
<div>
  <div>
    <u>
      _
    </u>
     Ha'i, from South Pacific
  </div>
</div>
`;

exports[`Highlighting for different types of clues 13`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    "What's a Grecian ___?" "About $
    <span
      class="emotion-0"
    >
      3 a
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      2-down Then...
      <b
        class="emotion-2"
      >
        [
        S
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
     week!"
  </div>
</div>
`;
